osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs

```diff
@@ -142,23 +161,33 @@ public async Task JoinRoom(Room room)
         /// <returns>The joined <see cref="MultiplayerRoom"/>.</returns>
         protected abstract Task<MultiplayerRoom> JoinRoom(long roomId);
 
-        public virtual Task LeaveRoom()
+        public Task LeaveRoom()
         {
-            Scheduler.Add(() =>
+            // Leaving rooms is expected to occur instantaneously whilst the operation is finalised in the background.
+            // However a few members need to be reset immediately to prevent other components from entering invalid states whilst the operation hasn't yet completed.
+            // For example, if a room was left and the user immediately pressed the "create room" button, then the user could be taken into the lobby if the value of Room is not reset in time.
+            var scheduledReset = scheduleAsync(() =>
             {
-                if (Room == null)
-                    return;
+                // The join may have not completed yet, so certain tasks that either update the room or reference the room should be cancelled.
+                // This includes the setting of Room itself along with the initial update of the room settings on join.
+                joinCancellationSource?.Cancel();
```

Comment:
This sounds correct.. @smoogipoo can you check this and confirm intention?

Reply:
The room updates from `JoinRoom` are not necessarily scheduled already.

Suppose a `LeaveRoom` happens that schedules before the schedule on [L144](https://github.com/smoogipoo/osu/blob/90ba8ae234bbec60aab5b42f9cd625eb8e2d2f02/osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs#L140-L144) of `JoinRoom`. In this case the room will be set to null by the `LeaveRoom` schedule but then set to non-null by a previous invocation of `JoinRoom`.

As for whether this could be unscheduled... I think so.